compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -369,9 +369,22 @@ pub(super) fn mir_assign_valid_types<'tcx>(
         // needs layout equality. For performance reason we skip this check when
         // the types are equal. Equal types *can* have different layouts when
         // enum downcast is involved (as enum variants carry the type of the
-        // enum), but those should never occur in assignments.
+        // enum), but those should never occur in assignments. Except in the
+        // one case where it does happen: different valid scalar ranges on two
+        // otherwise equal layouts; for those we do the layout comparison while
+        // ignoring the specific ranges. This comes up for example if you have
+        // a newtype struct tagged with `#[rustc_layout_scalar_valid_range_start(..)]`.
         if cfg!(debug_assertions) || src.ty != dest.ty {
-            assert_eq!(src.layout, dest.layout);
+            if let (Abi::Scalar(_), Abi::Scalar(_)) = (src.layout.abi(), dest.layout.abi()) {
+                assert!(
+                    src.equal_up_to_abi(&dest),
+                    "layout doesn't match. src={:?}, dest={:?}",
+                    src.layout,
+                    dest.layout
+                );
+            } else {
+                assert_eq!(src.layout, dest.layout);
+            }
```

Comment:
Miri being affected by what is effectively a codegen optimization makes me feel even more strongly that this "effective validity range" should be in `PlaceRef` next to the (unoptimized) `TyAndLayout`.

Reply:
`ScalarRanges` seems a bit unfortunate but yeah that was the general idea, the diff is small enough that I'm not really worried about the details either way!